src/main/java/com/google/devtools/build/lib/concurrent/MultiExecutorQueueVisitor.java

```diff
@@ -0,0 +1,125 @@
+      ExecutorService executionPhaseExecutorService,
+      boolean failFastOnException,
+      ErrorClassifier errorClassifier) {
+    super(
+        regularPoolExecutorService,
+        /*shutdownOnCompletion=*/ true,
+        failFastOnException,
+        errorClassifier);
+    this.regularPoolExecutorService = super.getExecutorService();
+    this.cpuHeavyPoolExecutorService = Preconditions.checkNotNull(cpuHeavyPoolExecutorService);
+    this.executionPhaseExecutorService = executionPhaseExecutorService;
+  }
+
+  public static MultiExecutorQueueVisitor createWithExecutorServices(
+      ExecutorService regularPoolExecutorService,
+      ExecutorService cpuHeavyPoolExecutorService,
+      boolean failFastOnException,
+      ErrorClassifier errorClassifier) {
+    return createWithExecutorServices(
+        regularPoolExecutorService,
+        cpuHeavyPoolExecutorService,
+        /*executionPhaseExecutorService=*/ null,
+        failFastOnException,
+        errorClassifier);
+  }
+
+  public static MultiExecutorQueueVisitor createWithExecutorServices(
+      ExecutorService regularPoolExecutorService,
+      ExecutorService cpuHeavyPoolExecutorService,
+      ExecutorService executionPhaseExecutorService,
+      boolean failFastOnException,
+      ErrorClassifier errorClassifier) {
+    return new MultiExecutorQueueVisitor(
+        regularPoolExecutorService,
+        cpuHeavyPoolExecutorService,
+        executionPhaseExecutorService,
+        failFastOnException,
+        errorClassifier);
+  }
+
+  @Override
+  public final void execute(Runnable runnable, ThreadPoolType threadPoolType) {
+    super.executeWithExecutorService(runnable, getExecutorServiceByThreadPoolType(threadPoolType));
+  }
+
+  @VisibleForTesting
+  ExecutorService getExecutorServiceByThreadPoolType(ThreadPoolType threadPoolType) {
+    switch (threadPoolType) {
+      case REGULAR:
+        return regularPoolExecutorService;
+      case CPU_HEAVY:
+        return cpuHeavyPoolExecutorService;
+      case EXECUTION_PHASE:
+        Preconditions.checkNotNull(executionPhaseExecutorService);
+        return executionPhaseExecutorService;
+    }
+    throw new IllegalStateException("Invalid ThreadPoolType: " + threadPoolType);
+  }
+
+  @Override
+  protected void shutdownExecutorService(Throwable catastrophe) {
+    if (catastrophe != null) {
+      Throwables.throwIfUnchecked(catastrophe);
+    }
+    internalShutdownExecutorService(regularPoolExecutorService);
+    internalShutdownExecutorService(cpuHeavyPoolExecutorService);
+    if (executionPhaseExecutorService != null) {
+      internalShutdownExecutorService(executionPhaseExecutorService);
+    }
+  }
+
+  private void internalShutdownExecutorService(ExecutorService executorService) {
+    executorService.shutdown();
+    while (true) {
+      try {
+        executorService.awaitTermination(Integer.MAX_VALUE, SECONDS);
+        break;
+      } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)